### DISCUSSION TITLE
关于M9A版本同步的问题

### DISCUSSION THREAD

**Discussion by MLAcookie** [MLAcookie/MAA1999WPF#14] [Q&A]
我想问下各位怎么给M9A做版本更新好
我现在的想法是写一个VersionHelper之类的，本身也可以独立出一个Cli工具
用来检测M9A的release更新情况（调用Github的API），有新的也可以下载
版本的检测就靠interface.json最下面的version字段

这样的话后续做应用本身检测M9A的更新也解决了
（我实在是没想明白怎么用项目引用的方式来解决这个问题

**Comment by 1m188** [comment]
可以查看一下MSBuild的官方文档，里面有对于自定义task的说明，可以用代码自定义某些构建过程：
https://learn.microsoft.com/zh-cn/visualstudio/msbuild/msbuild-tasks?view=vs-2022
https://learn.microsoft.com/zh-cn/visualstudio/msbuild/task-writing?view=vs-2022

我最近也在看这个，说不定能有帮助。

关于项目引用，我的想法是，通过某种手段让被依赖项目的构建结果输出只是一个包含M9A的文件夹，然后M9AWPF依赖该项目的构建结果，就会在M9AWPF构建前先构建该项目，并把该项目构建结果复制到M9AWPF的构建目录下，就像是依赖一个输出为dll的项目时的行为一样。

**Reply by MLAcookie** [reply]
看了一天，感觉不是我这几天能解决的😓，我还是先做M9A的更新吧，这个快解决了

**Comment by KaronGH** [comment]
我赞成MLA佬的方式，如果只是在Build的过程中添加Task的话，对于只使用Release的用户而言是无法进行自动更新的
我简单检索了一下，现在的初步想法是独立于M9AWPF之外添加一个M9AWPF.VersionHelper Project，然后编译的时候可以编译到同一个Release目录内（MM9AWPF.exe平级）这样不仅能够方便解决网络问题，也方便做CLI


不过如果做双项目的话，当前的目录结构就需要修改一下，把M9AWPF放到子目录（项目）下面，我可以先试着做一版PR看看，如果不需要的话可以再Close

**Reply by MLAcookie** [reply]
现在的话先解决用户在GUI里面能更新就可以了
CLI这些东西也就我们开发者会用，实在不行的话也就是手动复制一下的事情
之前188大佬写的build脚本也差不多解决了开发时M9A版本的问题

不过如果说M9A更新工具能一块把GUI的版本一起更新了，那也确实是需要这样的双项目结构的
这个也不着急，感觉别的功能还有界面之类的还能再打磨打磨

**Comment by MLAcookie** [comment]
如果 #17 没啥问题的话我就合并主干了

**Reply by KaronGH** [reply]
Merge吧，不过我还有一些想法
1. 开发的时候其实是不需要自动处理M9A版本更新的，因为能自己Build项目的肯定知道报错的话会去下M9A，不需要构建自动下载。自动下载功能更多的是给用户设计的
2. 现在多项目设计是依据master分支进行的，所以主分支合并的代码我会修改后添加到多项目分支，多项目的分支会直接提交到dev-addUpdater，如果有想法也可以直接到这个分支下面看
